A87-grpc-metrics-xds-outlier-detection.md

```diff
@@ -0,0 +1,65 @@
+A91: gRPC Metrics for Outlier Detection
+---
+* Author(s): @pardhukonakanchi, @huntsman90
+* Approver: @markdroth
+* Status: In Review
+* Implemented in: Core, Go
+* Last updated: 2025-03-05
+* Discussion at: https://groups.google.com/g/grpc-io/c/iMezDbq5U-g
+
+## Abstract
+
+This document proposes some new metrics that will be added in gRPC for xDS Client Outlier Detection
+
+## Background
+
+[A50: gRPC xDS Outlier Detection Support](https://github.com/grpc/proposal/blob/master/A50-xds-outlier-detection.md) is a spec for gRPC to support Outlier Detection. The current implementation only offers debug and trace logging in terms of visibility, which can be insufficient for understanding and diagnosing decision making in large-scale production systems. 
+
+Using A79: Non-per-call Metrics Architecture, it is possible to add granular metrics to make visibility into outlier detection easy for service owners utilizing gRPC.
+
+### Related proposals: 
+* [A50: gRPC xDS Outlier Detection Support](https://github.com/grpc/proposal/blob/master/A50-xds-outlier-detection.md)
+* [A66: OpenTelemetry Metrics](https://github.com/grpc/proposal/blob/master/A66-otel-stats.md)
+* [A79: Non-per-call Metrics Architecture](https://github.com/grpc/proposal/blob/master/A79-non-per-call-metrics-architecture.md)
+
+## Proposal
+
+[A79’s](https://github.com/grpc/proposal/blob/master/A79-non-per-call-metrics-architecture.md) non-per-call metrics architecture fits perfectly into the metrics reporting solution. We propose to aim for parity where appropriate with [Envoy’s Outlier Detection metric collection](https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#outlier-detection-statistics) to be collected in gRPC
+
+Outlier Detection metrics will have the following labels:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | Indicates the target we are running outlier detection on. (Same as the attribute defined in [A66](https://github.com/grpc/proposal/blob/master/A66-otel-stats.md).) |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+|  grpc.lb.outlier_detection.ejections_enforced_total | Counter | {ejection} | 	grpc.target |	Total enforced ejections due to any outlier type |
+|  grpc.lb.outlier_detection.ejections_active | Gauge |	{ejection} |	grpc.target |	Number of currently ejected hosts |
+|  grpc.lb.outlier_detection.ejections_overflow |	Counter |	{ejection} |	grpc.target |	Number of ejections aborted due to max ejection percentage |
+|  grpc.lb.outlier_detection.ejections_enforced_success_rate |	Counter |	{ejection} |	grpc.target |	Enforced success rate outlier ejections |
+|  grpc.lb.outlier_detection.ejections_detected_success_rate |	Counter |	{ejection} |	grpc.target |	Detected (but unenforced) success rate outlier ejections |
```

Comment:
The name of this metric seems to include all detected success rate ejections, including the ones that were not enforced, but the description seems to indicate that it only includes unenforced ejections. Which of those is intended? `grpc.lb.outlier_detection.ejections_detected_failure_percentage` also has the same issue.